throughput/pacer.go

```diff
@@ -6,57 +6,80 @@ package throughput
 import (
 	"context"
 	"fmt"
+	"sync"
 
 	"github.com/ava-labs/hypersdk/api/ws"
 	"github.com/ava-labs/hypersdk/chain"
-	"github.com/ava-labs/hypersdk/utils"
 )
 
 type pacer struct {
 	ws *ws.WebSocketClient
 
 	inflight chan struct{}
-	done     chan error
+	done     chan struct{}
+
+	errOnce sync.Once
+	err     error
 }
 
 func (p *pacer) Run(ctx context.Context, max int) {
 	p.inflight = make(chan struct{}, max)
-	p.done = make(chan error, 1)
+	p.done = make(chan struct{}, 1)
 
-	// Start a goroutine to process transaction results
-	for range p.inflight {
-		txID, result, err := p.ws.ListenTx(ctx)
-		if err != nil {
-			p.done <- fmt.Errorf("error listening to tx %s: %w", txID, err)
-			return
-		}
-		if result == nil {
-			p.done <- fmt.Errorf("tx %s expired", txID)
-			return
-		}
-		if !result.Success {
-			// Should never happen
-			p.done <- fmt.Errorf("tx failure %w: %s", ErrTxFailed, result.Error)
+	for {
+		select {
+		case _, ok := <-p.inflight:
+			if !ok {
+				p.setError(nil)
+				return
+			}
+			txID, result, err := p.ws.ListenTx(ctx)
+			if err != nil {
+				p.setError(fmt.Errorf("error listening to tx %s: %w", txID, err))
+				return
+			}
+			if result == nil {
+				p.setError(fmt.Errorf("tx %s expired", txID))
+				return
+			}
+			if !result.Success {
+				// Should never happen
+				p.setError(fmt.Errorf("tx failure %w: %s", ErrTxFailed, result.Error))
+				return
+			}
+		case <-p.done:
 			return
 		}
 	}
-	p.done <- nil
 }
 
 func (p *pacer) Add(tx *chain.Transaction) error {
+	if p.err != nil {
+		return p.err
+	}
+
 	if err := p.ws.RegisterTx(tx); err != nil {
-		return err
+		p.setError(err)
+		return p.err
```

Comment:
Do we need to check `p.err` here or can we register the tx and read the err iff `p.done` has been closed?

Since this value is set at most once immediately before `p.done` is closed, we should technically check that the channel has been closed before reading this value.